src/main/java/br/ufpe/cin/mergers/SemistructuredMerge.java

```diff
@@ -107,167 +119,204 @@ public static String merge(File left, File base, File right, MergeContext contex
+		List<FSTNode> nonTerminalAChildren = nonTerminalA.getChildren();
+
+		for (int i = 0; i < nonTerminalAChildren.size(); i++) {
+			FSTNode childA = nonTerminalAChildren.get(i);
+			FSTNode correspondentB = nonTerminalB.getCompatibleChild(childA);
+
+			if (correspondentB == null) { // is a new node from left, or a deleted base node in right
+				FSTNode cloneA = clone(nonTerminalA, childA);
 
-				/*
-				 * nodes from left or leftBase
-				 */
-				List<FSTNode> nonterminalAChildren = nonterminalA.getChildren();
-				
-				for (int i = 0; i < nonterminalAChildren.size(); i++) {
-					FSTNode childA = nonterminalAChildren.get(i);
-					FSTNode childB = nonterminalB.getCompatibleChild(childA);
-					
-					if (childB == null) { 								// is a new node from left, or a deleted base node in right
-						FSTNode cloneA = childA.getDeepClone();
-						if (childA.index == -1)
-							childA.index = nodeA.index;
-						cloneA.index = childA.index;
-
-						FSTNode childALeftNeighbour = getLeftNeighbourNode(nonterminalAChildren, i);
-						FSTNode childARightNeighbour = getRightNeighbourNode(nonterminalAChildren, i);
-						addNodeToNonTerminalNearNeighbour(cloneA, childALeftNeighbour, childARightNeighbour, nonterminalComposed);
-
-						if (context.deletedBaseNodes.contains(childA)) { // this is only possible when processing right nodes because this is a base node not present either in left and right
-							context.deletedBaseNodes.remove(childA);
-							context.deletedBaseNodes.add(cloneA);
-						}
-
-						if(isProcessingBaseTree){ //node added by left in relation to base
-							context.addedLeftNodes.add(cloneA);
-						} else {
-							if(!context.addedLeftNodes.contains(childA))
-								context.nodesDeletedByRight.add(cloneA);
-						}
-					} else {
-						if (!isProcessingBaseTree) {
-							context.deletedBaseNodes.remove(childA); 	// node common to right and base but not to left
-						}
+				FSTNode childALeftNeighbour = getLeftNeighbourNode(nonTerminalAChildren, i);
+				FSTNode childARightNeighbour = getRightNeighbourNode(nonTerminalAChildren, i);
+				addNodeToNonTerminalNearNeighbour(cloneA, childALeftNeighbour, childARightNeighbour, result);
+
+				if (step == SuperimpositionStep.Left_Base) { // node added by left in relation to base
+					context.addedLeftNodes.add(cloneA);
+				} else if (!context.addedLeftNodes.contains(childA)) {
+					context.nodesDeletedByRight.add(cloneA);
+
+					if (context.nodesDeletedByLeft.contains(cloneA)) {
+						context.deletedBaseNodes.add(cloneA);
 					}
 				}
-				return nonterminalComposed;
+			}
+		}
+	}
 
-			} else if (nodeA instanceof FSTTerminal && nodeB instanceof FSTTerminal	&& parent instanceof FSTNonTerminal) {
-				FSTTerminal terminalA = (FSTTerminal) nodeA;
-				FSTTerminal terminalB = (FSTTerminal) nodeB;
-				FSTTerminal terminalComposed = (FSTTerminal) composed;
+	private static void checkAddedNodes(FSTNonTerminal nonTerminalA, FSTNonTerminal nonTerminalB, MergeContext context,
```

Comment:
@jvcoutinho can you add comments explaining the check methods? and maybe choose names that better reveal what is being done?

Reply:
@jvcoutinho ?